WIP narrow: Add backend support for new direct message narrows.

laurynmm opened the pull request:
Adds backend support for fetching messages with the new direct message narrows:
- `"is:dm"` *(deprecates "is:private")*
- `"dm"` *(deprecates "pm-with")*
- `"dm-including"` *(deprecates "group-pm-with")*

Part of work on #24806.

See relevant CZO conversations in [api design](https://chat.zulip.org/#narrow/stream/378-api-design/topic/direct.20message.20search.20operators/near/1516511) and [design](https://chat.zulip.org/#narrow/stream/101-design/topic/.22Private.22.20vs.20.22Direct.22/near/1516241) streams.

<details>
<summary>git-grep of "pm-with" in /zerver after changes</summary>

```console
$ git grep pm-with zerver/
zerver/lib/narrow.py:            # Both "pm-with" and "group-pm-with" are deprecated operators.
zerver/lib/narrow.py:            "pm-with": self.by_dm,
zerver/lib/narrow.py:            "group-pm-with": self.by_group_pm_with,
zerver/lib/narrow.py:            operators_supporting_id = ["sender", "group-pm-with", "stream", "dm-including"]
zerver/lib/narrow.py:            operators_supporting_ids = ["pm-with", "dm"]
zerver/lib/url_encoding.py:    base_url = f"{realm.uri}/#narrow/pm-with/"
zerver/lib/url_encoding.py:    base_url = f"{realm.uri}/#narrow/pm-with/"
zerver/lib/url_encoding.py:        "pm-with",
zerver/tests/test_email_notifications.py:            f"view it in Zulip Dev Zulip: http://zulip.testserver/#narrow/pm-with/{cordelia.id}-{encoded_name}"
zerver/tests/test_message_fetch.py:    # Test that deprecated "pm-with" works.
zerver/tests/test_message_fetch.py:        term = dict(operator="pm-with", operand=self.hamlet_email)
zerver/tests/test_message_fetch.py:    # Test that the underscore version of deprecated "pm-with" works.
zerver/tests/test_message_fetch.py:    # Test that deprecated "group-pm-with" works.
zerver/tests/test_message_fetch.py:        term = dict(operator="group-pm-with", operand=self.othello_email)
zerver/tests/test_message_fetch.py:        term = dict(operator="group-pm-with", operand=self.othello_email, negated=True)
zerver/tests/test_message_fetch.py:        term = dict(operator="group-pm-with", operand="non-existing@zulip.com")
zerver/tests/test_message_fetch.py:    # Test that the underscore version of deprecated "group-pm-with" works.
zerver/tests/test_message_fetch.py:            is_spectator_compatible([{"operator": "pm-with", "operand": "hamlet@zulip.com"}])
zerver/tests/test_message_fetch.py:            is_spectator_compatible([{"operator": "group-pm-with", "operand": "hamlet@zulip.com"}])
zerver/tests/test_message_fetch.py:        A request for old messages with a narrow by deprecated "group-pm-with"
zerver/tests/test_message_fetch.py:            narrow = [dict(operator="group-pm-with", operand=operand)]
zerver/tests/test_message_fetch.py:        narrow = [dict(operator="group-pm-with", operand=cordelia.email)]
zerver/tests/test_message_fetch.py:        # and deprecated "group-pm-with" operators
zerver/tests/test_message_fetch.py:        for operand in ["sender", "group-pm-with", "stream", "dm-including"]:
zerver/tests/test_message_fetch.py:        self.assertEqual(url, "http://zulip.testserver/#narrow/pm-with/77,80-pm/near/555")
```
</details>

**Notes**:
- From the git-grep above, the only instances of "pm-with" that haven't been updated by these changes are those that directly reference `pm-with` URLs.
- Have done some manual testing fetching messages via curl and via dev tools in the web app, but next steps will be to implement support for these in the frontend/web app.
- For the `dm:self_user` case, in `add_term` where we build the query for sqlachemy, I used the same query that we use for `is:dm` since we basically want to return all of the logged in user's direct messages (group and 1:1) and added a comment to document why we do this.
- There's a prep commit before the "dm-including" implementation to share some code between it and "group-pm-with" where we get the recipient IDs for group direct messages. This commit could be moved to be anywhere before the "dm-including" commit in the series.
- Because the API documentation for the specific narrows we support is also part of the help center documentation, we will need to be careful with when we make those changes, which is why they are not included yet in these changes.
  - One thought I had was doing the API changelog feature level update and `version.py` bump with these changes.
  - And then updating the help center [search for messages](https://zulip.com/help/search-for-messages) and the API docs [construct narrow](https://zulip.com/api/construct-narrow) articles when the users can actually use these narrows in the web app (and other clients).

---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

laurynmm commented:
**Updates**:
- Added two commits for frontend implementation of new narrows.
  - First implemented `is:dm` to deprecate `is:private`.
    - When a user types `is:p` ... `is:private`, they will still get a search suggestion for `is:dm`. But the suggestions with user pills for `dm:` searches will not be filled in. But for the new `is:dm`, this works as expected.
  - Then implemented `dm:` to deprecate `pm-with:`.
    - I used the "from" alias for "sender" as a guide for my changes.
    - When a user types `pm-with:`, they will get search suggestions as if for `dm:`
- Existing links in messages (like quote and reply near links) or that exist in general should still work with these changes. The backend still returns messages for those inputs, but the web app updates the operator/operand when creating the new `Filter`.
  - The URL bar in the browser does not update with these changes for these preexisting links.

<details>
<summary>Some example links to use in basic testing</summary>

*I've gone through more thorough manual testing, but these are some examples. I'd also suggest prepping some near links via quote and reply to direct message conversations in the dev environment to make testing faster.*
```
http://localhost:9991/#narrow/is/private
http://localhost:9991/#narrow/is/private/-pm-with/10-King-Hamlet
http://localhost:9991/#narrow/is/starred/is/private
http://localhost:9991/#narrow/pm-with/8-Cordelia,-Lear's-daughter
http://localhost:9991/#narrow/pm-with/11-Iago/has/image
http://localhost:9991/#narrow/pm-with/9-Desdemona/near/10
```
</details>

**Still to-do**:
- Implement `dm-including` on the frontend.
- Update backend webhook `pm-with` near URLs.
- Add documentation and API feature level update, planned as final commit.

alya commented:
I played around with the PR, and the UX feels good to me!

laurynmm commented:
**Updates**:
- Added two new commits from previous update note:
  - A backend commit that updates the url-encoding, which is used for email notifications and outgoing webhooks, to use `dm` instead of `pm-with`.
  - A frontend commit that implements `dm-including`. This mirrors most of the implementation for `group-pm-with`.

---

**Still to-do**:
- Update `group-pm-with` for deprecated behavior. I feel like it will be cleaner to do this in a separate commit from the `dm-including` implementation. Also, it would be good to clarify on CZO what we expect the web app to do for `group-pm-with` moving forward.
- The final documentation commit with the API changelog / feature level bump as well as API, help center and `?` menu documentation updates.

---

**Screenshots for `dm-including`**:

<details>
<summary>Empty narrow banners - no direct messages</summary>

![Screenshot from 2023-04-17 20-07-50](https://user-images.githubusercontent.com/63245456/232572867-2103e2a2-d45f-4023-8a4d-e3de415d525f.png)
![Screenshot from 2023-04-17 19-50-22](https://user-images.githubusercontent.com/63245456/232572434-d9f4f8e8-b9dd-4226-a206-44499c571e1d.png)
</details>
<details>
<summary>Negated dm-including with current user - returns all stream messages</summary>

![Screenshot from 2023-04-17 19-51-58](https://user-images.githubusercontent.com/63245456/232572438-51121ebd-486b-4896-b954-4ec28d5dc29c.png)
</details>
<details>
<summary>dm-including with current user - returns all private messages</summary>

### Note: Like `group-pm-with` this view does not mark messages as read
![Screenshot from 2023-04-17 19-53-44](https://user-images.githubusercontent.com/63245456/232572442-24bf5559-3d80-466f-8729-a195a3ff15a7.png)
</details>
<details>
<summary>dm-including returns both group and 1-on-1 direct messages</summary>

![Screenshot from 2023-04-17 19-54-15](https://user-images.githubusercontent.com/63245456/232572449-8b3d2a30-2915-4643-af25-bad4f916fd6a.png)
![Screenshot from 2023-04-17 19-54-31](https://user-images.githubusercontent.com/63245456/232572453-c4578ad6-66d2-4349-9cdc-8d7867cce087.png)
![Screenshot from 2023-04-17 19-56-12](https://user-images.githubusercontent.com/63245456/232572455-a830492c-1926-408d-bcd5-0866bb151650.png)
</details>
<details>
<summary>Negated `dm-including` with another user combined with `is:dm` returns all direct messages without that user</summary>

![Screenshot from 2023-04-17 19-56-46](https://user-images.githubusercontent.com/63245456/232572463-4873aaed-c70e-4af4-9720-5767152d8f37.png)
</details>

timabbott commented on zerver/lib/narrow.py:
```diff
@@ -157,6 +157,12 @@ def narrow_filter(event: Mapping[str, Any]) -> bool:
                 if operand.lower() != message["sender_email"].lower():
                     return False
             elif operator == "is" and operand == "private":
+                # "private" is a deprecated operand with the "is" operator.
```
Somehow I feel like this comment would be more readable if we just said "# is:private is a legacy alias for is:dm".

timabbott commented on zerver/lib/narrow.py:
```diff
@@ -336,7 +342,13 @@ def by_is(self, query: Select, operand: str, maybe_negate: ConditionTransform) -
+            # For backwards-compatibility with links/URLs, we currently
+            # maintain support for fetching messages with these narrows.
             cond = column("flags", Integer).op("&")(UserMessage.flags.is_private.mask) != 0
             return query.where(maybe_negate(cond))
```
Would it make sense to write this as `operand in ["dm", "private"]`? I mainly ask because it's harder to mess up updating that code by forgetting that there's names for this operator.

(The query applies to both hunks in this file).

laurynmm replied:
Yeah, that's much easier to read and maintain. Updated as such.

timabbott commented on zerver/lib/narrow.py:
```diff
@@ -273,14 +273,18 @@ def __init__(
             "search": self.by_search,
-            "pm-with": self.by_pm_with,
+            "dm": self.by_dm,
+            # "pm-with" is a deprecated operator.
```
Again, I think saying "pm-with: is a deprecated alias of dm:" would make this a lot easier to follow for someone who doesn't already know how this works.

laurynmm replied:
Done.

timabbott commented on zerver/lib/narrow.py:
```diff
@@ -597,6 +598,24 @@ def by_dm(
+            ).values("recipient_id")
+        ]
+
+        return set(self_recipient_ids) & set(narrow_recipient_ids)
```
This should probably be left as a follow-up since it's orthogonal to the other changes here and will require fiddling with the query planner to make sure we're doing it right, but I'm almost certain the database queries here can be optimized by avoiding the list comprehensions that convert these into lists and letting the database layer do the final filtering...

laurynmm replied:
Yeah, these changes are big and complex enough that I feel like this as a separate follow-up task/issue makes sense.

timabbott commented on zerver/tests/test_message_fetch.py:
```diff
@@ -1826,10 +1857,107 @@ def test_get_visible_messages_with_narrow_dm(self) -> None:
+                me,
+                [cordelia, othello],
+            ),
+        )
```
It would be good to include a message to a huddle from someone other than `me`; only including messages sent by yourself to a huddle is a class of bug I can imagine in this code path.

laurynmm replied:
Expanded these test cases for other senders of huddle messages.

timabbott commented on web/src/filter.js:
```diff
@@ -222,6 +222,12 @@ export class Filter {
 
         switch (operator) {
+            case "is":
+                // deprecated "is:private" narrow
```
I would write "// is:private was renamed to is:dm" here -- that concisely explains the concrete reason why this is correct.

laurynmm replied:
Done.

timabbott commented on web/src/search_suggestion.js:
```diff
@@ -626,6 +626,14 @@ function get_sent_by_me_suggestions(last, operators) {
 }
 
 function get_operator_suggestions(last) {
+    // Offer suggestion for deprecated "is:private" filter
```
Better would be "// Suggest is:dm to anyone with is:private in their finger memory", since that explains the purpose of this logic a bit more precisely.

laurynmm replied:
I used 'muscle memory' since that seemed like a more generic / known term.

timabbott commented on web/tests/search_suggestion_now.test.js:
```diff
@@ -127,26 +127,38 @@ test("private_suggestions", () => {
     assert.deepEqual(suggestions.strings, expected);
 
+    // deprecated "is:private" still suggests "is:dm",
+    // but not user suggestions.
```
I'm not sure what "user suggestions" means here (and in the other place in this file).

laurynmm replied:
Updated to read:

```js
    // "is:private" was renamed to "is:dm", so
    // we suggest "is:dm" to anyone with "is:private"
    // in their muscle memory.
    query = "is:pr";
    suggestions = get_suggestions("", query);
    expected = ["is:dm"];
    assert.deepEqual(suggestions.strings, expected);

    query = "is:private";
    suggestions = get_suggestions("", query);
    expected = ["is:dm"];
    assert.deepEqual(suggestions.strings, expected);
```

timabbott commented:
I commented on the changes so far; generally seems pretty clean.

laurynmm commented on web/e2e-tests/message-basics.test.ts:
```diff
@@ -73,10 +73,10 @@ async function expect_huddle(page: Page): Promise<void> {
 }
 
-async function expect_cordelia_private_narrow(page: Page): Promise<void> {
+async function expect_cordelia_dm_narrow(page: Page): Promise<void> {
```
Makes sense. Updated this and a few other cases of this in the frontend commits.

laurynmm commented:
**Updates**:
- Addressed initial review feedback on first 7 commits (see individual comments above). Also, did a general pass to no longer use "DM" in any comments / code.
- Added a commit removing "group-pm-with" support from the web app and making it a legacy alias for "dm-including".
- Added the final commit with the API, help center and `?` menu documentation updates.

**Relevant CZO conversations**:
- [#api design > direct messages search operators](https://chat.zulip.org/#narrow/stream/378-api-design/topic/direct.20message.20search.20operators/near/1540897)
- [#api documentation > direct messages search operators](https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/direct.20message.20search.20operators/near/1551983)
- [#design > "Private" vs "Direct"](https://chat.zulip.org/#narrow/stream/101-design/topic/.22Private.22.20vs.20.22Direct.22/near/1529625)

---

**Documentation screenshots**:

<details>
<summary>Search for messages</summary>

[Current documentation](https://zulip.com/help/search-for-messages#search-by-location)
![Screenshot from 2023-04-20 15-01-14](https://user-images.githubusercontent.com/63245456/233381761-b2822acf-6d7f-48d6-a090-7316e7b0ff7b.png)
</details>
<details>
<summary>Web app ? menu</summary>

| Before | After |
| --- | --- |
|  ![Screenshot from 2023-04-20 15-30-40](https://user-images.githubusercontent.com/63245456/233382011-676f4fda-14e9-47ef-8939-73feffb9d0fc.png) | ![Screenshot from 2023-04-20 15-01-47](https://user-images.githubusercontent.com/63245456/233382009-06f3cc01-170b-4bda-84e7-eb5ec54572d7.png)
</details>
<details>
<summary>API changelog update</summary>

![Screenshot from 2023-04-20 14-59-52](https://user-images.githubusercontent.com/63245456/233382384-a956b876-3667-4454-a2cc-1547bc8cd5b9.png)
</details>
<details>
<summary>Construct narrow</summary>

[Current documentation](https://zulip.com/api/construct-narrow)
![Screenshot from 2023-04-20 15-00-14](https://user-images.githubusercontent.com/63245456/233382671-1fe10768-1ede-49b7-a23d-42b6d1817622.png)
![Screenshot from 2023-04-20 15-38-54](https://user-images.githubusercontent.com/63245456/233384007-0bc21f56-706e-4801-8e24-492eac4ca7a1.png)

</details>
<details>
<summary>Get messages</summary>

[Current documentation](https://zulip.com/api/get-messages#parameter-narrow)
![Screenshot from 2023-04-20 15-00-29](https://user-images.githubusercontent.com/63245456/233382928-583ee385-60fb-4d79-abb9-1a83b4000a10.png)
</details>
<details>
<summary>Register queue - narrow parameter</summary>

[Current documentation](https://zulip.com/api/register-queue#parameter-narrow)
![Screenshot from 2023-04-20 15-00-38](https://user-images.githubusercontent.com/63245456/233382938-220cd538-1f19-4248-8056-9a97db223556.png)
</details>
<details>
<summary>Check messages match narrow</summary>

[Current documentation](https://zulip.com/api/check-messages-match-narrow)
![Screenshot from 2023-04-20 15-00-45](https://user-images.githubusercontent.com/63245456/233382940-92acf1e0-4aaf-4320-99b9-6758cfd1858f.png)
</details>
<details>
<summary>Update personal message flags for narrow</summary>

[Current documentation](https://zulip.com/api/update-message-flags-for-narrow#parameter-narrow)
![Screenshot from 2023-04-20 15-00-51](https://user-images.githubusercontent.com/63245456/233382948-4ac0dba1-04d8-4c7a-9c2a-60fcf6a1a9c3.png)
</details>

alya commented:
The help center and `?` menu documentation looks good to me.

gnprice commented on api_docs/construct-narrow.md:
```diff
@@ -45,6 +45,13 @@ operator, search the current user's personal message history. See
 [searching shared history](/help/search-for-messages#searching-shared-history)
 for details.
 
+**Changes**: In Zulip 7.0 (feature level 177), support was added for
+  three narrows related to direct messages: `is:dm`, `dm` and
+  `dm-including`. The `dm` operator replaced and deprecated the
```
nit: I think of a "narrow" as meaning specifically a complete set of filters one might send to /api/get-messages. It looks like this page currently consistently uses "narrow" with that meaning too.

So `is:dm` is a narrow, but more generally it's a filter (an operator-operand pair) one can include in a narrow. And `dm` and `dm-including` are operators, which are building blocks of narrows but not narrows on their own.

If it were just `dm` and `dm-including`, I'd say "two operators". For all three of these, I think "three filters" would be fine. (Strictly it's three *kinds* of filters: `is:dm` plus `-is:dm`, and filters with the operators `dm` and `dm-including`. But that's more verbose and probably not helpful.)

laurynmm replied:
This makes sense. Updated version uses "filter" in general and "operator" for sentences that are about `dm` or `dm-including` specifically.

gnprice commented on api_docs/changelog.md:
```diff
@@ -20,6 +20,20 @@ format used by the Zulip server that they are interacting with.
+  [`GET /messages/matches_narrow`](/api/check-messages-match-narrow),
+  [`POST /message/flags/narrow`](/api/update-message-flags-for-narrow),
+  [`POST /register`](/api/register-queue):
+  Added support for three [search/narrows](/api/construct-narrow)
+  related to direct messages: `is:dm`, `dm` and `dm-including`.
```
nit: I'm not sure how to parse the phrase "three search/narrows related to direct messages". Is "searches/narrows" what you intended?

Per the other comment, though, I'd probably go for something like "three search/narrow filters", linking the phrase "search/narrow filters".

laurynmm replied:
I intended to use the same text as the [Changes in Zulip 2.1](https://zulip.com/api/changelog#changes-in-zulip-21) note, but obviously it got mangled as I edited/revised. Updated for your suggestion.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -5278,8 +5278,13 @@ paths:
-            **Changes**: In Zulip 2.1.0, added support for using user/stream IDs
-            when constructing narrows for a message's sender, its stream and/or
+            **Changes**: In Zulip 7.0 (feature level 177), support was
+            added for three narrows related to direct messages: `is:dm`,
```
similarly, s/three narrows/three filters/, here and a few places below

laurynmm replied:
Done. Also made sure the parameter description included "filter" so that the changes note would read clearly.

gnprice reviewed (commented):
Thanks! A few wording comments — really all variations on one comment.

timabbott commented on api_docs/construct-narrow.md:
```diff
@@ -45,6 +45,13 @@ operator, search the current user's personal message history. See
+  `dm-including`. The `dm` operator replaced and deprecated the
+  `pm-with` operator. The `is:dm` narrow replaced and deprecated the
+  `is:private` narrow. The `dm-including` operator replaced and
+  deprecated the `group-pm-with` operator.
```
I think this is the place where I'd spell out that `group-pm-with` had slightly different semantics from `dm-including` -- since this is the detailed documentation page one should use when one is interested in precise search operator semantics.

laurynmm replied:
Yes, this makes sense. Expanded this comment in the updated version.

timabbott commented:
I had one thing to add, the screenshots otherwise lgtm.

laurynmm commented:
@gnprice and @timabbott - Thanks for the feedback on the API documentation commit. I've updated for all the above comments. See screenshots below.

**Notes**:
- In the API changelog note, I still used "operator" when specifically referring to the two operator changes (`dm` and `dm-including`). But in the general case or with the `is:dm` / `is:private` case I used "filter". 
- The [construct a narrow]()  **Changes** note has been expanded to cover the difference between `dm-including` and `group-pm-with`. I also added a specific note that the other two filters pairs (new/deprecated) return the same results.
- The **Changes** note for the four endpoint parameters is the same, so I have only one example screenshot below. I did add "filter" to a few of the narrow parameter descriptions so that the **Changes** note would read more clearly.

---

**Updated screenshots**:
<details>
<summary>Revised API changelog note</summary>

![Screenshot from 2023-04-21 13-18-55](https://user-images.githubusercontent.com/63245456/233626309-6e0a8631-f91d-48c2-b5f0-9779906c4845.png)
</details>
<details>
<summary>Construct a narrow - expanded Changes note</summary>

![Screenshot from 2023-04-21 13-21-52](https://user-images.githubusercontent.com/63245456/233626331-b91c4a3f-6f2f-4efb-b590-e413f6f07c06.png)
</details>
<details>
<summary>Updated Changes note in endpoint parameter</summary>

![Screenshot from 2023-04-21 13-19-50](https://user-images.githubusercontent.com/63245456/233626351-67ed3e18-e9ee-4f54-b278-0c62301d30f8.png)
</details>

laurynmm commented on web/src/filter.js:
```diff
@@ -868,7 +883,9 @@ export class Filter {
+                case "dm":
                 case "pm-with":
                 case "sender":
                 case "from":
```
I didn't delete either "pm-with" or "group-pm-with" here because the "from" operator was here, but I think all three could be removed after these changes because it's always called on a filter that's been set. And all three are updated in `Filter.canonicalize_term` when the filter is initialized.

laurynmm commented:
Failing test is likely a [puppeteer flake](https://chat.zulip.org/#narrow/stream/43-automated-testing/topic/main.20failing/near/1552708) in `copy-and-paste.test.ts`.

laurynmm commented:
**Updates**:
- See [above comment](https://github.com/zulip/zulip/pull/25003#issuecomment-1517702940) about changes to API documentation per review feedback.
- The final commit notes that this PR will close #24806. The new direct message filters (`is:dm`, `dm` and `dm-including`) have the updated URLs. I feel like updating the URLs for existing/legacy links/searches should be a separate follow-up issue. I've updated the original comment to note that on this PR as well.
- I had one note/question ([see above](https://github.com/zulip/zulip/pull/25003#discussion_r1173761612) about `Filter.update_email`, but I believe that could also be a small clean-up done after these changes.

gnprice commented on api_docs/construct-narrow.md:
```diff
@@ -45,6 +45,24 @@ operator, search the current user's personal message history. See
+the `is:private` filter. The `dm-including` operator replaced and
+deprecated the `group-pm-with` operator.
+
+The `dm-including`and `group-pm-with` operators return slightly
```
nit:
```suggestion
The `dm-including` and `group-pm-with` operators return slightly
```

laurynmm replied:
Fixed.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -5259,27 +5259,30 @@ paths:
-            streams, so bots using this API should either be
-            subscribed to appropriate streams or use a shared history
-            search narrow with this endpoint.
+            Newly created bot users are not usually subscribed to any streams,so
```
nit:
```suggestion
            Newly created bot users are not usually subscribed to any streams, so
```

gnprice replied:
But actually better: this paragraph isn't changing at all except for rewrapping, right? Probably better to just leave it untouched, then. Same for the two paragraphs before it.

laurynmm replied:
Yeah, instead I'm making a note to come back and clean up some of the line wrapping later.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -5259,27 +5259,30 @@ paths:
         - name: narrow
           in: query
           description: |
-            The narrow where you want to fetch the messages from. See how to
+            The narrow filter where you want to fetch the messages from. See how to
             [construct a narrow](/api/construct-narrow).
```
I think this is actually clearer as just "narrow". (Similarly the other two `narrow` parameters below.) I think of the whole thing as "a narrow", consisting of a list of filters.

Calling the whole thing "a narrow" is also consistent with the other text here, and with the parameter's name `narrow`.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -5259,27 +5259,30 @@ paths:
 
-            **Changes**: In Zulip 2.1.0, added support for using user/stream IDs
-            when constructing narrows for a message's sender, its stream and/or
+            **Changes**: In Zulip 7.0 (feature level 177), support was added for
+            three filters related to direct messages: `is:dm`, `dm` and `dm-including`;
```
If you want to contextualize the term "filter" here, some options would be:

"three narrow filters"
(or "three narrow-filters", to avoid sounding like filters that are narrow)

"support in narrows was added for three filters"

"narrows gained support for three new filters"

laurynmm replied:
Thanks for these suggestions. I went with "narrows gained support for three new filters".

gnprice reviewed (commented):
Thanks! A bit more on the meaning of "narrow" vs. "filter".

timabbott commented:
Tagging as a release goal since we're certainly including this in 7.0 at this point; please re-add integration review when Greg's comments are addressed.

laurynmm commented:
**Updates**:
- Addressed @gnprice's feedback above.
  - The endpoint parameter **Changes** notes contextualize the use of "filter".
  - Fixed the missing space in `construct_narrow.md` and removed the line wrapping edits/clean up (with a note to myself to revisit as a separate refactor/clean up PR).

**Updated screenshots**:

<details>
<summary>Construct narrow changes note</summary>

![Screenshot from 2023-04-25 15-15-47](https://user-images.githubusercontent.com/63245456/234290771-a29013ec-5a5f-4b7b-9600-5df5c33632f2.png)

</details>
<details>
<summary>Example of endpoint `narrow` parameter changes note</summary>

![Screenshot from 2023-04-25 15-16-05](https://user-images.githubusercontent.com/63245456/234290531-b185b591-9e6f-47c6-8d82-4859257c004e.png)

</details>

timabbott commented:
This is great, merged, thanks for executing this complex API transition @laurynmm!

Brett2148 reviewed (commented):
[![image](https://github.com/user-attachments/assets/c27236f4-994a-4e86-aacb-9cf4580982f5)]()
